### PR TITLE
handle sub-repo permissions filtering in git.DiffPath

### PR DESCRIPTION
**Pull request by mollylogue**
- Add sub-repo permissions filtering to `git.DiffPath`
- Add tests for `git.DiffPath`
NOTE: should be merged after https://github.com/sourcegraph/sourcegraph/pull/29213
Closes https://github.com/sourcegraph/sourcegraph/issues/28912

### CODE REVIEW

**Review by sashaostrikov — APPROVED** [review]
LGTM!

**Review by efritz — APPROVED** [review]

File: internal/vcs/git/diff.go [hunk 1]
```diff
@@ -63,7 +64,10 @@ func Diff(ctx context.Context, opts DiffOptions) (*DiffFileIterator, error) {
 
 // DiffPath returns a position-ordered slice of changes (additions or deletions)
 // of the given path between the given source and target commits.
-func DiffPath(ctx context.Context, repo api.RepoName, sourceCommit, targetCommit, path string) ([]*diff.Hunk, error) {
+func DiffPath(ctx context.Context, repo api.RepoName, sourceCommit, targetCommit, path string, checker authz.SubRepoPermissionChecker) ([]*diff.Hunk, error) {
+	if hasAccess, err := authz.HasAccessToPath(ctx, checker, repo, path); err != nil || !hasAccess {
+		return nil, err
```
Comment (ryanslade) [inline internal/vcs/git/diff.go:69]: Related the comment from the other PR. I think of they don't have access we should return an `os.ErrNotExist` error.

**Review by ryanslade — APPROVED** [review]

File: internal/vcs/git/diff.go (hunk 1 above)
Comment (mollylogue) [reply internal/vcs/git/diff.go:69]: @ryanslade Which other PR? I'm looking through my other open PRs and I'm not seeing this comment but definitely could be missing something 🤔

Comment (ryanslade) [reply internal/vcs/git/diff.go:70]: Sorry, should have linked to it 😄  

https://github.com/sourcegraph/sourcegraph/pull/29213

**Pull request merged**